Provide an xcversion action to select xcode via version specifier

oysta opened the pull request:
This implements #4414

There are some caveats/open questions:
- This breaks backwards compatibility with the previous `xcode_select` action as it used a peculiar (for Fastlane) non-keyworded argument to specify the path. This patch mandates either `path:` or `version:` be used. I saw no easy way to preserve compatibility.
- I've introduced a dependency on the `xcode-install` gem in the tests. I'm not sure if there is a better way to handle this.
- This is the first Ruby I've written for production consumption in a fair while. There are likely problems with it and better ways to do things.

oysta committed:
Allow xcode_select to use a version requirement

oysta committed:
Fix hound style violations

oysta committed:
Fix multi-line assignment indenting

KrauseFx commented on fastlane/fastlane.gemspec:
```diff
@@ -65,4 +65,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'appium_lib', '~> 4.1.0'
   spec.add_development_dependency 'rest-client', '~> 1.6.7'
   spec.add_development_dependency 'fakefs', '~> 0.8.1'
+  spec.add_development_dependency 'xcode-install'
```
I'm not 100% up to date if we can do this, let's ask @neonichu. We've had some conflicts in the past. If it works now, I'm totally okay with adding this dependency

KrauseFx replied:
This needs to specify the version we need too 👍

neonichu replied:
Should be fine now, since there's no longer any pinning on an older version of `spaceship` present.

![](http://i.imgur.com/c4jt321.png)

KrauseFx commented on fastlane/lib/fastlane/actions/xcode_select.rb:
```diff
@@ -18,32 +18,92 @@ module Actions
 
-        # Verify that a param was passed in
-        UI.user_error!("Path to Xcode application required (e.x. \"/Applications/Xcode.app\")") unless xcode_path.to_s.length > 0
+      def self.find_xcode(req)
```
I'm not a big fan with adding new methods here, could we move this to a helper class? Similar to `crashlytics_helper.rb`

KrauseFx commented on fastlane/lib/fastlane/actions/xcode_select.rb:
```diff
@@ -18,32 +18,92 @@ module Actions
+        end
+      rescue Gem::LoadError
+        UI.error("The 'xcode-install' gem is needed to select Xcode by version number.")
+        Actions.verify_gem!('xcode-install') # This is awkward and destined to fail but is used to get the install help messages
```
This was designed to be called on top of the action to tell the user how to install the gem. Any reason why we don't just add it on top of `find_xcode`?

oysta replied:
@KrauseFx The only reason the code is arranged awkwardly like this was so that I could add the "The 'xcode-install' gem is needed to select Xcode by version number." message to make it clear why it was necessary. As the gem is only needed in when `version:` is specified I thought it may have been confusing to just error out with the `verify_gem!` message. Happy to bin it if you feel it is unnecessary.

KrauseFx commented on fastlane/lib/fastlane/actions/xcode_select.rb:
```diff
@@ -18,32 +18,92 @@ module Actions
         [:ios, :mac].include? platform
       end
+
+      module Verify
```
I would love to also have this in a helper class instead of in the action.

KrauseFx commented on fastlane/lib/fastlane/actions/xcode_select.rb:
```diff
@@ -18,32 +18,92 @@ module Actions
 
+      def self.available_options
+        [
+          FastlaneCore::ConfigItem.new(key: :path,
```
Why would a user specify the `:path`? You could also use the `xcode_select` action, right?

oysta replied:
@KrauseFx This is the `xcode_select` action. I had to introduce a `path` keyword to support that behaviour. This is the part I'm most uncomfortable with because it breaks backwards compatibility with the existing version. That is:

```ruby
xcode_select("/Applications/Xcode.app")
```

would have to be updated to be:

```ruby
xcode_select(path: "/Applications/Xcode.app")
```

KrauseFx commented:
This is a great pull request, thanks for your great contribution. I added some comments. I like how we use the existing RubyGems system to check the dependency. 

I'm curios if this should be part of `xcode-install` instead of `fastlane`, as it might be useful for the CI too? Would love to hear what @neonichu thinks.

oysta commented:
Thanks @KrauseFx. I'm very happy to help. I'll adjust the code for your feedback; probably have something ready tomorrow. 👍

neonichu commented:
`xcode-install` already has the `select` command, but that changes the Xcode version system-wide. A `DEVELOPER_DIR` based approach wouldn't really work, because it only subprocesses would inherit it. So I think it makes perfect sense inside `fastlane`, as it can control the environment for its subprocesses.

KrauseFx commented:
Very cool, thanks everyone. @oysta I don't think we need the `verify_gem` call, since we have `xcode-install` as a dependency anyway

oysta commented:
Unless I've missed something, I don't believe `xcode-install` is a direct dependency for fastlane at the moment. Users don't have to install it unless they are going to use a part of fastlane that requires it. In this case they only need it if the wan't to select by version; selecting by path does not require it. That's why I have the `verify_gem` call. There's another `verify_gem` call for it in the `xcode_install` action at the moment. I take it that this is the desired behaviour? Or do we want to make `xcode_install` a non-optional dependency and remove these calls throughout?

I've listed it as a _development_ only dependency in the gemspec because it is needed for the tests.

neonichu commented:
I think it was made optional because of the conflicts due to different versions of spaceship being used? So it might be possible to just depend on it again now.

KrauseFx commented:
This pull request now adds the dependency https://github.com/fastlane/fastlane/pull/4421/files#diff-0a40ff76d18a73ce31894adf33b04b0fR68

neonichu commented:
But only as a development dependency.

oysta committed:
Move action helper methods into helper class

oysta committed:
Make xcode-install a direct gem dependency

Remove verify_gem! calls from xcode_select action because of the new
direct dependency.

oysta committed:
Remove verify_gem call from xcode_install action

The xcode-install gem is now a direct dependency.

oysta commented:
I think I've addressed all of the feedback up to now. I'd still like to know if it is okay that this will actually break all existing uses of `xcode_select` because of the new requirement for keywords to be used.

KrauseFx commented:
> I'd still like to know if it is okay that this will actually break all existing uses of xcode_select because of the new requirement for keywords to be used.

We can't introduce breaking changes with actions, since it breaks existing setups. We have to still support the old way of calling the action.

KrauseFx commented:
Looking more into this, let's make this a separate action, it makes more sense to still provide the simple class that just sets an environment variable, and to provide this one, that does more things. The documentation should be very clear on how these 2 are different

oysta commented:
> Why is it breaking? Because before this change people would just call the action using xcode_select "/Applications/Xcode.app"? 

Yes — now they have to use the `path:` keyword. Ideally I'd be able to hook both calling styles, but it appears I can't do this easily.

> If so, I'd suggest maybe just having a new action for this with a different name, how about xcversion, which is the name of the executable of xcode-install

This makes sense; I can make the change. As a slight counter-argument, it does leave open the possibility of both actions being used and treading on each other.

oysta committed:
Merge remote-tracking branch 'upstream/master' into feature/xcode-select-by-version

oysta committed:
Move Xcode version selection to xcversion action

This reverts the changes to the xcode_select action, restoring it to its
original behaviour.

oysta commented:
I've moved the logic to a new `xcversion` action. `xcode_select` has been completely reverted to the original version.

KrauseFx commented on fastlane/lib/fastlane/helper/xcversion_helper.rb:
```diff
@@ -0,0 +1,26 @@
+
+      module Verify
+        def self.requirement(req)
+          puts "SFDDSFS" + req
```
Is that debug code?

oysta replied:
urgh. Sorry.

KrauseFx commented on fastlane/lib/fastlane/helper/xcversion_helper.rb:
```diff
@@ -0,0 +1,26 @@
+      module Verify
+        def self.requirement(req)
+          puts "SFDDSFS" + req
+          UI.user_error!("version must be specified") if req.nil? || req.to_s.strip.size == 0
```
Let's capitalize: `Version must be ...`

KrauseFx replied:
Don't forget to also update the tests

KrauseFx commented:
This is great, thank you so much for updating the PR. I prefer this style 👍

KrauseFx commented:
I added 2 small comments before it's ready to be merged 👍

oysta committed:
Remove debug code

oysta committed:
Capitalise 'Version' in error message

oysta commented:
All done

oysta committed:
Update podspec dependency comment

Since xcode_select was reverted, this now applies to the xcversion
action.

KrauseFx commented:
This is great, thanks for updating the PR 👍 This will be in the next release
